flask_security/twofactor.py

```diff
@@ -0,0 +1,162 @@
+
+# Convenient references
+_security = LocalProxy(lambda: app.extensions['security'])
+
+_datastore = LocalProxy(lambda: _security.datastore)
+
+
+def send_security_token(user, method, totp_secret):
+    """Sends the security token via email for the specified user.
+    :param user: The user to send the code to
+    :param method: The method in which the code will be sent
+                ('mail' or 'sms') at the moment
+    :param totp_secret: a unique shared secret of the user
+    """
+    token_to_be_sent = get_totp_password(totp_secret)
+    if method == 'mail':
+        send_mail(config_value('EMAIL_SUBJECT_TWO_FACTOR'),
+                  user.email,
+                  'two_factor_instructions',
+                  user=user,
+                  token=token_to_be_sent)
+    elif method == 'sms':
+        msg = "Use this code to log in: %s" % token_to_be_sent
+        from_number = config_value('TWO_FACTOR_SMS_SERVICE_CONFIG')[
+            'PHONE_NUMBER']
+        if 'phone_number' in session:
+            to_number = session['phone_number']
+        else:
+            to_number = user.phone_number
+        sms_sender = SmsSenderFactory.createSender(
+            config_value('TWO_FACTOR_SMS_SERVICE'))
+        sms_sender.send_sms(from_number=from_number,
+                            to_number=to_number, msg=msg)
+
+    elif method == 'google_authenticator':
+        # password are generated automatically in the google authenticator app
+        pass
+
+
+def get_totp_uri(username, totp_secret):
+    """ Generate provisioning url for use with the qrcode
+            scanner built into the app
+    :param username: username of the current user
+    :param totp_secret: a unique shared secret of the user
+    :return:
+    """
+    service_name = config_value('TWO_FACTOR_URI_SERVICE_NAME')
+
+    return 'otpauth://totp/{0}:{1}?secret={2}&issuer={0}'\
+        .format(service_name, username, totp_secret)
+
+
+def verify_totp(token, totp_secret, window=0):
+    """ Verifies token for specific user_totp
+    :param token - token to be check against user's secret
+    :param totp_secret - a unique shared secret of the user
+    :param window - optional, compensate for clock skew, number of
+        intervals to check on each side of the current time.
+        (default is 0 - only check the current clock time)
+    :return:
+    """
+    return onetimepass.valid_totp(token, totp_secret, window=window)
+
+
+def get_totp_password(totp_secret):
+    """Get time-based one-time password on the basis of given secret and time
+    :param totp_secret - a unique shared secret of the user
+    """
+    return onetimepass.get_totp(totp_secret)
+
+
+def generate_totp():
+    return base64.b32encode(os.urandom(10)).decode('utf-8')
+
+
+def generate_qrcode():
+    """generate the qrcode for the two-factor authentication process"""
+    if 'google_authenticator' not in\
+            config_value('TWO_FACTOR_ENABLED_METHODS'):
+        return abort(404)
```

Comment:
In general I think that utility functions like this should be assuming a request context - rather throw an exception to the caller?